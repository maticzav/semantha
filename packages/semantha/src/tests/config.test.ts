import * as path from 'path'
import { getConfiguration } from '..'

describe('configuration', () => {
  test('getConfiguration reports missing repository configuration', async () => {
    const cwd = path.resolve(__dirname, './__fixtures__/config/no-repository/')
    const config = await getConfiguration(cwd)

    expect(config).toEqual({
      status: 'err',
      message: 'Missing workspaces or repository definition in package.json',
    })
  })

  test('getConfiguration reports missing workspaces configuration', async () => {
    const cwd = path.resolve(__dirname, './__fixtures__/config/no-workspaces/')
    const config = await getConfiguration(cwd)

    expect(config).toEqual({
      status: 'err',
      message: 'Missing workspaces or repository definition in package.json',
    })
  })

  test('getConfiguration reports misconfigured repository', async () => {
    const cwd = path.resolve(__dirname, './__fixtures__/config/invalid/')
    const config = await getConfiguration(cwd)

    expect(config).toEqual({
      status: 'err',
      message: `Couldn't parse provided repository faulty repo`,
    })
  })

  test('getConfiguration correctly handles errors', async () => {
    const cwd = path.resolve(__dirname, './__fixtures__/config/whatever/')
    const workspaces = await getConfiguration(cwd)

    /* Tests */

    expect(workspaces).toEqual({
      status: 'err',
      message: `ENOENT: no such file or directory, open '${cwd}/package.json'`,
    })
  })

  test('getConfiguration errors on no-workspace matches', async () => {
    const cwd = path.resolve(__dirname, './__fixtures__/config/no-matches/')
    const workspaces = await getConfiguration(cwd)

    /* Tests */

    expect(workspaces).toEqual({
      status: 'err',
      message: "Couldn't find any workspace.",
    })
  })

  test('getConfiguration loads correct configuration', async () => {
    const cwd = path.resolve(__dirname, './__fixtures__/config/valid/')
    const config = await getConfiguration(cwd)

    expect(config).toEqual({
      config: {
        repository: {
          owner: 'maticzav',
          repo: 'semantha',
        },
        workspaces: [
          path.resolve(
            __dirname,
            './__fixtures__/config/valid/packages/package-a',
          ),
          path.resolve(
            __dirname,
            './__fixtures__/config/valid/packages/package-b',
          ),
        ],
      },
      status: 'ok',
    })
  })
})
