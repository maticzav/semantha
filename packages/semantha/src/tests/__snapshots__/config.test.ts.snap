// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`configuration getConfigurationFrom finds correct configuration 1`] = `
Object {
  "config": Object {
    "repository": Object {
      "owner": "maticzav",
      "repo": "maticzav/semantha",
    },
    "workspaces": Array [
      Object {
        "path": "/Users/maticzavadlal/Code/sandbox/semantha/packages/semantha/src/tests/__fixtures__/workspaces/valid/packages/package-a",
        "pkg": Object {
          "dependencies": Object {
            "package-b": "1.0.2",
            "something": "1.0.0",
          },
          "devDependencies": Object {
            "devit": "1.0.0",
            "package-c": "2.0.0",
          },
          "name": "package-a",
          "version": "1.0.0",
        },
      },
      Object {
        "path": "/Users/maticzavadlal/Code/sandbox/semantha/packages/semantha/src/tests/__fixtures__/workspaces/valid/packages/package-b",
        "pkg": Object {
          "dependencies": Object {
            "package-c": "2.0.0",
            "something": "1.0.0",
          },
          "devDependencies": Object {
            "devit": "1.0.0",
          },
          "name": "package-b",
          "version": "1.0.2",
        },
      },
      Object {
        "path": "/Users/maticzavadlal/Code/sandbox/semantha/packages/semantha/src/tests/__fixtures__/workspaces/valid/packages/package-c",
        "pkg": Object {
          "dependencies": Object {
            "something": "1.0.0",
          },
          "devDependencies": Object {
            "devit": "1.0.0",
          },
          "name": "package-c",
          "version": "1.0.0",
        },
      },
    ],
  },
  "status": "ok",
}
`;
